project/BuildPlugin.scala

```diff
@@ -0,0 +1,359 @@
+  import BuildImplementation.{ BuildDefaults, BuildResolvers }
+
+  val baseVersion: String = "1.1.0-SNAPSHOT"
+  val internalPath: File = file("internal")
+  val bridgeScalaVersions: List[String] =
+    List(Dependencies.scala212, Dependencies.scala211, Dependencies.scala210)
+
+  val ZincGitHomepage: URL = url("https://github.com/sbt/zinc")
+  val ScalaCenterMaintainer: Developer =
+    Developer("jvican", "Jorge Vicente Cantero", "@jvican", url("https://github.com/jvican"))
+
+  // Ids that we will use to name our projects in build.sbt
+  val CompilerInterfaceId = "compiler-interface"
+  val CompilerBridgeId = "compiler-bridge"
+  val ZincApiInfoId = "zinc-apiinfo"
+
+  // Defines the constants for the alternative publishing
+  val ZincAlternativeCacheName = "alternative-local"
+  val ZincAlternativeCacheDir: File = file(sys.props("user.home") + "/.ivy2/zinc-alternative")
+
+  // Defines several settings that are exposed to the projects definition in build.sbt
+  private[this] val noPublishSettings: Seq[Def.Setting[_]] = BuildDefaults.noPublishSettings
+  def noPublish(p: Project): Project = p.settings(noPublishSettings)
+
+  val adaptOptionsForOldScalaVersions: Seq[Def.Setting[_]] =
+    List(Keys.scalacOptions := BuildDefaults.zincScalacOptionsRedefinition.value)
+  // Sets up mima settings for modules that have to be binary compatible with Zinc 1.0.0
+  val mimaSettings: Seq[Def.Setting[_]] =
+    List(MimaKeys.mimaPreviousArtifacts := BuildDefaults.zincPreviousArtifacts.value)
+
+  import sbt.{ TaskKey, taskKey }
+  val scriptedPublishAll = taskKey[Unit]("Publishes all the Zinc artifacts for scripted")
+  val cleanSbtBridge: TaskKey[Unit] = taskKey[Unit]("Cleans the sbt bridge.")
+  val zincPublishLocal: TaskKey[Unit] =
+    taskKey[Unit]("Publishes Zinc artifacts to a alternative local cache.")
+  val zincPublishLocalSettings: Seq[Def.Setting[_]] = List(
+    Keys.resolvers += BuildResolvers.AlternativeLocalResolver,
+    zincPublishLocal := BuildDefaults.zincPublishLocal.value,
+  )
+
+  val tearDownBenchmarkResources: TaskKey[Unit] = taskKey[Unit]("Remove benchmark resources.")
+  val benchmarksTestDir = sbt.IO.createTemporaryDirectory
+
+  val sourcesForAllScalaVersionsSetting: Seq[Def.Setting[_]] =
+    List(Keys.unmanagedSourceDirectories ++= BuildDefaults.handleScalaSpecificSources.value)
+  def inCompileAndTest(ss: Def.Setting[_]): Seq[Def.Setting[_]] =
+    List(Compile, Test).flatMap(sbt.inConfig(_)(ss))
+}
+
+object BuildImplementation {
+  import sbt.{ ScmInfo }
+  val buildSettings: Seq[Def.Setting[_]] = List(
+    Scripted.scriptedBufferLog := true,
+    GitKeys.baseVersion := BuildKeys.baseVersion,
+    GitKeys.gitUncommittedChanges := BuildDefaults.gitUncommitedChanges.value,
+    BintrayKeys.bintrayPackage := "zinc",
+    ScalafmtKeys.scalafmtOnCompile := true,
+    ScalafmtKeys.scalafmtVersion := "1.2.0",
+    ScalafmtKeys.scalafmtOnCompile in Sbt := false,
+    Keys.description := "Incremental compiler of Scala",
```

Comment:
Keys are in a single shared namespace, which is why you can refer to them by name (e.g `TaskKey[Unit]("publishSigned") := (())`). So we should just import them instead of referring to them like this. Particularly when you end up with this repetition:

```scala
    GitKeys.gitUncommittedChanges := BuildDefaults.gitUncommitedChanges.value,
    BintrayKeys.bintrayPackage := "zinc",
    ScalafmtKeys.scalafmtOnCompile := true,
```

Could you revert back please?

Reply:
> Keys are in a single shared namespace,
  
Which conflicts with other things in scope because they have too many generic names, like `moduleID` et al. I don't want to accidentally be referring to a local variable named this way because it's a source of bugs.
  
The good things of having everything accessible by `Keys` are:

* It's clear that the keys I'm defining come from sbt (which is really clear for us if we don't use `Keys`, but not for people reading this build).
* I can do `val moduleID = Keys.moduleID.value`. Otherwise I would have to invent a similar name which doesn't clash with `moduleID`. I could choose between `id` (id of what? project? module?), `module` (which is not really true, it's an id of a module) or another variation, but this is *weird* and I don't want to spend time thinking how to name local variables that are straightforward to identify.
  
Does this make sense?

Reply:
> I don't want to accidentally be referring to a local variable named this way because it's a source of bugs.

Could you provide an example?

> It's clear that the keys I'm defining come from sbt

I don't think that's a strong enough reason to diverge in style.

> I can do `val moduleID = Keys.moduleID.value`

You can still do that with `import Keys._`.

---

Another reason not to go with this new style is that when we have slash syntax the build code and shell notation will be unified, so we'll want to use in both cases:

```scala
Compile / javacOptions
```

not

```scala
Compile / Keys.javacOptions
```

Reply:
> You can still do that with import Keys._.
  
No, you cannot. If you do `import Keys._`, `moduleID` will be in scope, so if I define `val moduleID = moduleID.value` Scala will think it's a recursive definition and will fail both at compile-time and runtime (in case you ever make to typecheck).
  
> Could you provide an example?
  
Say I define a local method called `publish` that defines how a thing is going to be published. If I've done `import Keys._`, `publish` will resolve to different terms depending on the scope I use them, which is misleading a source of bugs. I don't want my code to change its meaning depending on the place I use them. Not only it's not refactor friendly, it's confusing.
  
The problem gets worse when you don't even know that you're importing a term from `Keys.scala` that is conflicting with a term you've defined. What about `artifact`, `doc`, `mainClass`, `run` and the other bazillion of names that are really likely to appear in my build logic and conflict with my definitions?
  
> Another reason not to go with this new style is that when we have slash syntax the build code and shell notation will be unified
  
That's a matter of style. I don't mind the second version. My arguments here have to do with semantics and user readability (in terms of not having name clashes -- when users read they don't need to think if a term is defined at the end of the file, in an enclosing scope or comes from sbt -- they have a clear idea that `Keys.artifact` means the setting key from sbt).

Reply:
Inside of an AutoPlugin definition it is customary to `import Keys._`. I tend to import any autoImports from plugins that I works on to emulate what one would have in `build.sbt`.

Reply:
Yes, it is customary @eed3si9n. But I believe that, based on my technical explanations before, this particular implementation is superior than the "customary bad practice". I don't see any technical benefit on making the build plugin implementation be more similar to `build.sbt`. Perhaps you can elaborate on that?

Reply:
You both asked me to remove this. I did, but in the other PR because rebasing otherwise was hell. https://github.com/sbt/zinc/pull/429
  
I think this is good to go.

Reply:
Dale's been looking at this PR more than me, so I'll defer my vote to him.
@dwijnand Is this PR good to go as is? if not could you make a list of pending items?